src/bgw/job.c

```diff
@@ -0,0 +1,250 @@
+{
+	Catalog    *catalog = catalog_get();
+	ScannerCtx	scanCtx = {
+		.table = catalog->tables[BGW_JOB].id,
+		.index = CATALOG_INDEX(catalog, BGW_JOB, indexid),
+		.nkeys = nkeys,
+		.scankey = scankey,
+		.tuple_found = tuple_found,
+		.filter = tuple_filter,
+		.data = data,
+		.lockmode = lockmode,
+		.scandirection = ForwardScanDirection,
+		.result_mctx = mctx,
+	};
+
+	return scanner_scan(&scanCtx);
+}
+
+static inline int
+bgw_job_scan_job_id(int32 bgw_job_id, tuple_found_func tuple_found, tuple_filter_func tuple_filter,
+					void *data, MemoryContext mctx, LOCKMODE lockmode)
+{
+	ScanKeyData scankey[1];
+
+	ScanKeyInit(&scankey[0],
+				Anum_bgw_job_stat_pkey_idx_job_id,
+				BTEqualStrategyNumber, F_INT4EQ, Int32GetDatum(bgw_job_id));
+	return bgw_job_scan(BGW_JOB_STAT_PKEY_IDX,
+						scankey, 1, tuple_found, tuple_filter, data, mctx, lockmode);
+}
+
+static bool
+bgw_job_tuple_found(TupleInfo *ti, void *const data)
+{
+	BgwJob	  **job_pp = data;
+
+	*job_pp = bgw_job_from_tuple(ti->tuple, ti->mctx);
+	return false;
+}
+
+BgwJob *
+ts_bgw_job_find(int32 bgw_job_id, MemoryContext mctx)
+{
+	BgwJob	   *job_stat = NULL;
+
+	bgw_job_scan_job_id(bgw_job_id, bgw_job_tuple_found, NULL, &job_stat, mctx, AccessShareLock);
+
+	return job_stat;
+}
+
+bool
+ts_bgw_job_execute(BgwJob *job)
+{
+	if (namestrcmp(&job->fd.job_type, VERSION_CHECK_JOB_NAME) == 0)
+	{
+		/* TODO: fill in when merging with Amy's stuff  */
+		return false;
+	}
+	else
+	{
+#ifdef DEBUG
+		if (strncmp(NameStr(job->fd.job_type), TEST_JOB_TYPE_PREFIX, strlen(TEST_JOB_TYPE_PREFIX)) == 0)
+		{
+			PGFunction test_function = load_external_function(extension_get_so_name(), TEST_JOB_FUNCTION, false, NULL);
+
+			if (test_function == NULL)
+				elog(ERROR, "cannot find test execution function");
+
+			return DatumGetBool(DirectFunctionCall1(test_function, Int32GetDatum(job->fd.id)));
+		}
+#endif
+
+		elog(ERROR, "unknown job type \"%s\"", NameStr(job->fd.job_type));
+	}
+}
+
+bool
+ts_bgw_job_has_timeout(BgwJob *job)
+{
+	Interval	zero_val = {.time = 0,};
+
+	return !DatumGetBool(DirectFunctionCall2(interval_le, IntervalPGetDatum(&job->fd.max_runtime), IntervalPGetDatum(&zero_val)));
+}
+
+TimestampTz
+ts_bgw_job_timeout_at(BgwJob *job, TimestampTz start_time)
+{
+	return DatumGetTimestampTz(DirectFunctionCall2(
+												   timestamptz_pl_interval,
+												   TimestampTzGetDatum(start_time),
+												   IntervalPGetDatum(&job->fd.max_runtime)));
+}
+
+extern Datum
+ts_bgw_job_entrypoint(PG_FUNCTION_ARGS)
+{
+	Oid			db_oid = DatumGetObjectId(MyBgworkerEntry->bgw_main_arg);
+	int32		job_id = Int32GetDatum(DirectFunctionCall1(int4in, CStringGetDatum(MyBgworkerEntry->bgw_extra)));
+	BgwJob	   *job;
+	bool		res = false;
+
+	BackgroundWorkerBlockSignals();
+	/* Setup any signal handlers here */
+	BackgroundWorkerUnblockSignals();
+
+	elog(DEBUG1, "started background job %d", job_id);
+
+	BackgroundWorkerInitializeConnectionByOid(db_oid, InvalidOid);
```

Comment:
Given that we've already connected to the DB as the scheduler, it seems highly unlikely it will have disappeared before we run this bit. In fact, I think it might be impossible, given that if you drop the db, then we'll stop the worker etc. We haven't yet addressed the potential problems with eventually wanting to be able to run jobs as not a superuser, but that's a different story for a different day.